Task2

NikolayPV opened the pull request.

sergeyratz committed:
Task 2 for student practice

Commit:
First model of answer
- realization algorithm

Commit:
chenged realization
added passing by reference
the answer display  step by step

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -0,0 +1,44 @@
+#include <bits/stdc++.h>
```
Нельзя так инклюдить все либы из std. Нужно инклюдить только те что тебе необходимы.
В данный момент тебе нужно заинклюдить только iostream.

NikolayPV replied:
Исправил

a-nechyporenko commented on src/Task2/main.cpp:
```diff
@@ -10,7 +10,7 @@
 using namespace std;
 
-void insertionSort1(int size, vector <int> arr);
+void insertionSort1(int size, vector <int> &arr);
```
Передавать количество элементов вектора вместе с вектором избыточно, контейнеры реализованы в std предоставляют интерфейсы для получения этой информации.

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -0,0 +1,44 @@
+
+using namespace std;
+
+void insertionSort1(int size, vector <int> &arr);
```
Название функции не должны называться с цифрами. Как ты отличишь `insertionSort1` от `insertionSort2`? Как понять какая между ними разница? Так же Антон писал в прошлый раз что не нужен `size`. Ты можешь это достать из самого вектора. Открываем [доку](http://www.cplusplus.com/reference/vector/vector/) и читаем, там есть все необходимые функции.
Ссылка это часть типа что ты передаешь. по этому не:
`vector<int> &arr`
А должно быть:
`vector<int>& arr`
Так же лучше называть полностью, ты не пират :)
Так что:
`vector<int>& array`

NikolayPV replied:
исправил

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -0,0 +1,44 @@
+
+int main() 
+{
+	unsigned int ARRAY_SIZE;
```
что-то не так с табуляцией, везде должна быть одинаковая. К тому же лучше её лучше заменить пробелами в кол-ве трех. Капсом через нижний прочерк называют только константы. Так что просто `arraySize`.

NikolayPV replied:
исправил

a-nechyporenko commented on src/Task2/main.cpp:
```diff
@@ -10,7 +10,7 @@
+void insertionSort1(int size, vector <int> &arr);
 
 int main() 
 {
```
возможно будет лучше вынести это в функцию, которая будет участвовать в инициализации вектора.

KyryloRud replied:
ввод числа - да, функция может делегировать ввод числа, но никак не вместе с инициализацией вектора, это нужно разделять. Банальная замена контейнера с vector на list тебе все поломает, хотя и там и там можно пользоваться схожими методами, типа `push_back()`

a-nechyporenko replied:
я имел ввиду что создавать перейменую типа unsigned int  длиной в жизнь программы это как-то не очень. а вот отделить функцию которая будет возвращать число которое было введено только для инициализации вектора. 
 И о листе я что-то не понял что замена может поломать? ( http://ru.cppreference.com/w/cpp/container/list/list )

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -0,0 +1,44 @@
+int main() 
+{
+	unsigned int ARRAY_SIZE;
+   cin >> ARRAY_SIZE;
```
Не безопасно вводишь, если пользователь введет цифру больше/меньше чем может уместить в себя int, либо букву или символ -> программа упадет в рантайме.

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -0,0 +1,44 @@
+   vector<int> array(ARRAY_SIZE);
+   for(int i = 0; i < ARRAY_SIZE; ++i)
+	{
+		cin >> array[i];
```
Снова ввод без проверки на валидность.

NikolayPV replied:
Да, это не является главной задачей. Поэтому Серый сказал не делать.

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -0,0 +1,44 @@
+		cin >> array[i];
+   }
+   insertionSort1(ARRAY_SIZE, array);
+   for(int j = 0; j < ARRAY_SIZE; ++j) 
```
лучше раскрой скобки фигурные. Если их нет и ты забудешь при расширении кода добавить их, то может быть не правильный флоу. Всегда ставь скобки. Тебе скинли Coding Style, обращайся к нему. Если не ошибаюсь там только с фигурными скобками.

NikolayPV replied:
исправил

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -0,0 +1,44 @@
+{
+	int cutedElement = arr.back();
+	int currentElemet = vectorSize - 2;
+	while(cutedElement <= arr.at(currentElemet))
```
добавь перенос строки перед циклом, это вкусовщина, лично мне так проще увидеть где заканчивается код и начинается цикл.

NikolayPV replied:
добавил

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -0,0 +1,44 @@
+	{
+		arr[currentElemet + 1] = arr[currentElemet];
+		--currentElemet;
+		for(int j = 0; j < vectorSize; ++j) 	
```
Снова цикл без явного скоупа. Добавь скобки пожалуйста.

NikolayPV replied:
исправил

KyryloRud reviewed (commented):
Needs to refactor.

a-nechyporenko commented on src/Task2/main.cpp:
```diff
@@ -22,24 +22,23 @@ int main()
-	} 
-//	for(int j = 0; j < vectorSize; ++j) 	// ���� ������� ����� � ���� ����� ��������� ��������������� ������. 
-//		cout << arr[j] << ' ';
+		arr[currentElemet + 1] = arr[currentElemet];
```
Почему здесь не использовать at?

KyryloRud replied:
По тому, что это не является необходимой операцией. По сути как и выше, от функции `at()` можно избавиться. [Читаем доку.](http://www.cplusplus.com/reference/vector/vector/at/) Она не нужна по тому, что единственное чем она отличается от `operator[]`, это проверка, является ли елемент с таким индексом в пределах массива и если нет - кидает исключение. В своем коде нужно стараться работать вообще без исключений. Все должно работать однозначно и очень редко когда можно пребигать к их использованию (к примеру когда выделяеться память и то не всегда, см. `new`). Кроме того, тут это только понизит performance (хоть по факту этого и не заметишь тут). В данно функции он не изменяет размер массива и может гетнуть его размер. **По этому я бы советовал использовать только **`operator[]`**.**

a-nechyporenko commented on src/Task2/main.cpp:
```diff
@@ -22,24 +22,23 @@ int main()
+		arr[currentElemet + 1] = arr[currentElemet];
+		--currentElemet;
+		for(int j = 0; j < vectorSize; ++j) 	
+		cout << arr[j] << ' ';
```
зачем разделять на две строчки?

KyryloRud replied:
он не разделяет на строчки, а выводит через пробелы, но все равно +1 Зачем?

a-nechyporenko commented on src/Task2/main.cpp:
```diff
@@ -22,24 +22,23 @@ int main()
    insertionSort1(ARRAY_SIZE, array);
-//   for(int j = 0; j < ARRAY_SIZE; ++j) // �� ������� ������ �� �� ������� �����.
-//		cout << array[j] << ' ';
+   for(int j = 0; j < ARRAY_SIZE; ++j) 
```
можно тоже вынести в функцию

KyryloRud replied:
вывод массива на екран?

Commit:
efficient algorithm
deleted variable vectorSize in function - insertionSort
code's format is changed

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -9,36 +9,46 @@
 
-void insertionSort1(int vectorSize, vector <int> &arr) 
+//-----------------------------------------------------------------------------------------
+void insertionSort (vector <int>& _array) 
```
1. Вектор, тип вектора и ссылка это все один тип, не нужно его разрывать.
2. Нет необходимости писать проччерк перед названием переменной, по кодинг стайлу, что тебе скинул @Olnayd тебе нужно называть переменные по другому.

> [3.1.4. Variable names must be in mixed case starting with lower case.](http://geosoft.no/development/cppstyle.html)

`vector<int>& array`

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -9,36 +9,46 @@
-	int currentElemet = vectorSize - 2;
-	while(cutedElement <= arr.at(currentElemet))
+	int cutedElement = _array.back();
+	int currentElemet = _array.size() - 2;
```
ты часто используешь по коду ниже функцию `size()`, вместо этого создай константную переменную, которая будет хранить размер массива `const int arraySize( array.size() );`

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -9,36 +9,46 @@
+			
+			for(int i = 0; i < _array.size(); ++i) 
+			{
+				cout << _array[i] << ' ';
```
Функция называется `insertionSort()` - из навания не видно, что она выводит что-то на екран (да и по факту это было бы странно, если каждый сортируемый массив выводился бы на экран). Вынеси вывод массива из функции. Идиологически тут он лишний. Кроме того вывод будет не отсортированного массива.

**P.S.**
Напиши если это так по заданию необходимо делать. По тому что если нет, то это **нужно убрать**.

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -9,36 +9,46 @@
+			}
+			cout << endl;
+		}
+		currentElemet < 0 ?	_array[0] = cutedElement: _array[currentElemet + 1] = cutedElement;
```
Краткий `if` тут не оправдан. Лучше напиши обыкновенным, что бы читабильность была проще. К слову, вместо пробела, в строке используется табуляция.

KyryloRud commented on src/Task2/main.cpp:
```diff
@@ -9,36 +9,46 @@
-   cin >> ARRAY_SIZE;
-   vector<int> array(ARRAY_SIZE);
-   for(int i = 0; i < ARRAY_SIZE; ++i)
+	unsigned int arraySize;
```
Символ табуляции вместо пробелов.

KyryloRud reviewed (commented):
# Need to rework.
Я скомпилил код, и он не работает корректно.
Компилятор: `gcc version 4.6.3`.
1. Входные данные:
5 (размер массива), 5, 4, 3, 2, 1.
2. Ожидаемые выходные данные:
1, 2, 3, 4, 5.
3. Реальные выходные данные:
5 4 3 2 2 
5 4 3 3 2 
5 4 4 3 2 
5 5 4 3 2 
1 5 4 3 2
### Запустить код можно тут:
https://repl.it/@KirillHelm/NikolayPVStudentPracticeTask2

a-nechyporenko commented on src/Task2/main.cpp:
```diff
@@ -9,36 +9,46 @@
+	int cutedElement = _array.back();
+	int currentElemet = _array.size() - 2;
+	
+	if(_array.size() > 1)
```
если я правильно понимаю это условие проверяет пустой ли контейнер тебе пришел. В контейнерах реализована функция для подобной проверки.

a-nechyporenko commented on src/Task2/main.cpp:
```diff
@@ -9,36 +9,46 @@
-   insertionSort1(ARRAY_SIZE, array);
-   for(int j = 0; j < ARRAY_SIZE; ++j) 
+   insertionSort(array);
+   for(int j = 0; j < arraySize; ++j) 
```
в данном случае конечно так можно написать, но вообще размер контейнера запрашивай у самого контейнера  или пользуйся итераторами. Например вдруг ты используешь функции а они где-то там изменяют размер твоего контейнера. Кстати можешь как раз почитать про краткую запись фор.

a-nechyporenko commented on src/Task2/main.cpp:
```diff
@@ -9,36 +9,46 @@
+		{
+			_array[currentElemet + 1] = _array[currentElemet];
+			
+			for(int i = 0; i < _array.size(); ++i) 
```
ты выводишь масив здесь и в мейне, почему бы этот функционал не вынести в функцию?
